nntrainer/dataset/raw_file_data_producer.cpp

```diff
@@ -0,0 +1,180 @@
+#include <nntrainer_error.h>
+#include <node_exporter.h>
+#include <util_func.h>
+
+namespace nntrainer {
+
+/**
+ * @brief Props containing file path value
+ *
+ */
+class PropsPath : public Property<std::string> {
+public:
+  static constexpr const char *key = "path"; /**< unique key to access */
+  using prop_tag = str_prop_tag;             /**< property type */
+
+  bool isValid(const std::string &v) {
+    std::ifstream file(v, std::ios::binary | std::ios::ate);
+    return file.good();
+  }
+
+  void set(const std::string &v) {
+    Property<std::string>::set(v);
+    std::ifstream file(v, std::ios::binary | std::ios::ate);
+    cached_pos_size = file.tellg();
+  }
+
+  std::ifstream::pos_type file_size() { return cached_pos_size; }
+
+private:
+  std::ifstream::pos_type cached_pos_size;
+};
+
+RawFileDataProducer::RawFileDataProducer() : raw_file_props(new PropTypes()) {}
+RawFileDataProducer::~RawFileDataProducer() {}
+
+const std::string RawFileDataProducer::getType() const {
+  return RawFileDataProducer::type;
+}
+
+unsigned long long
+RawFileDataProducer::size(const std::vector<TensorDim> &input_dims,
+                          const std::vector<TensorDim> &label_dims) const {
+  auto size_accumulator = [](const unsigned int &a, const TensorDim &b) {
+    return a + b.getFeatureLen();
+  };
+
+  auto sample_size =
+    std::accumulate(input_dims.begin(), input_dims.end(), 0u, size_accumulator);
+  sample_size = std::accumulate(label_dims.begin(), label_dims.end(),
+                                sample_size, size_accumulator);
+
+  auto path_prop = std::get<PropsPath>(*raw_file_props);
+  auto file_size = path_prop.file_size();
+
+  /// checking alignment is a good way to make check if a file is valid,
+  /// unfortunately, our test dataset does not have this property
+  /// (trainingSet.dat, valSet.dat, testSet.dat) after checking, we can
+  /// uncomment below line.
+  // NNTR_THROW_IF((file_size % sample_size * RawFileDataProducer::pixel_size !=
+  // 0),
+  //               std::invalid_argument)
+  //   << " Given file does not align with the given sample size, sample size: "
+  //   << sample_size << " file_size: " << file_size;
+
+  return path_prop.file_size() /
+         (sample_size * RawFileDataProducer::pixel_size);
+}
+
+void RawFileDataProducer::setProperty(
+  const std::vector<std::string> &properties) {
+  auto left = loadProperties(properties, *raw_file_props);
+  NNTR_THROW_IF(!left.empty(), std::invalid_argument)
+    << "There is unparsed properties, size: " << left.size();
+}
+
+DataProducer::Generator
+RawFileDataProducer::finalize(const std::vector<TensorDim> &input_dims,
+                              const std::vector<TensorDim> &label_dims) {
+
+  /****************** Validation ****************/
+  auto sz = size(input_dims, label_dims);
+  auto batch = input_dims[0].batch();
+
+  NNTR_THROW_IF(sz < batch, std::invalid_argument)
+    << "calculated sample size is less than a batch";
+
```

Comment:
I'm not sure which one is better. but how about change the batch size with warrning, instead of giving error.

Reply:
As per the offline discussion two weeks ago, this will change to a element-wise generator so this validation will go away, I'll add todo here and apply this on the way to the migration :)